### ISSUE TITLE
callbackFiller removed in 5.0.0

### ISSUE THREAD

**Issue by Maelstrom96** [node-cache-manager/node-cache-manager#417]
I would like to know if there was a particular reason why this feature was removed? I might be misunderstanding the current behavior of `.wrap()` in v5.0.0, but by looking at the code, it seems like there isn't any mechanism left to coalesce simultaneous cache wrap for the same key, unlike before with `callbackFiller`. 

If I'm not mistaken and there isn't indeed a mechanism currently that does that, would you be open of a PR that re-introduce a `callbackFiller`-like feature? We were about to update before noticing this, which would really affect our application with how we've expected node-cache-manager to work, locking us out of the v5.0.0 with the await/async and Typescript support.

**Comment by Maelstrom96** [comment]
`Promise.all` has real world advantages when it comes to IO wait or DB query operations, where you are waiting and not doing anything - It's super easy to demonstrate too when tracing a request that uses `Promise.all` VS a normal for-each loop. With an await for-each, the execution would wait for a database query to resolve before processing with the next one.

In our situations, the database/cache queries are inside classes, with each class objects sometime requiring data that is shared with other objects. The previous `callbackFiller` function handled that use case perfectly for us and would coalesce those identical `cache.wrap()`.

I agree that in compute-only cases, `Promise.all` actually slows things down with no real benefits, since you won't get more CPU cycles to do those compute tasks, and `Promise.all` will just add more overhead.

So I understand you're not interested in re-implementing this and we should look at implementing this outside of the caching library?

**Comment by botika** [comment]
> So I understand you're not interested in re-implementing this and we should look at implementing this outside of the caching library?

Yes, exally. I don't want this features in my nodejs backends. 

https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/Rust/ntex/src/main_plt.rs#L34 

As you can see there is a single socket entry per thread in the form of an `rx` and `tx` tuple. 

The flow of this future is I try to read from the socket doing a poll, if it is multipart I keep reading, if the reading is OK I write data to the socket. So between reading and writing, is your logic. 
I.e. `Read -> **Your Logic** -> Write` only one at a time per thread. 

So whatever you put in between will be launched sequentially because it's a single thread. 
At most it will vectorize some operations.

**Comment by jonapgar-groupby** [comment]
What does Rust multithreading have to do with this issue?

Unhelpful responses IMO

**Comment by botika** [comment]
It's to explain that only one call is made at a time per thread. 

Therefore, if a parallel call occurs, it is because it has been called like this, calling the same function twice does not make sense in the same call. 
The only way to match 2 equal queries to the database is if they are called from different threads. This can't be since nodejs is a single threaded runtime.

**Comment by Maelstrom96** [comment]
> It's to explain that only one call is made at a time per thread.

But Rust is not NodeJS, is still doesn't make sense to use that as an example.

> Therefore, if a parallel call occurs, it is because it has been called like this, calling the same function twice does not make sense in the same call. The only way to match 2 equal queries to the database is if they are called from different threads. This can't be since nodejs is a single threaded runtime.

That is simply not true. NodeJS has indeed a single threaded event loop, but the asynchronous nature of NodeJS allows it to process other things while waiting for an event/callback. That means that it is 100% possible to process multiple database queries at the (almost) same time while the previous one(s) is still not resolved.

That is a trace on my current NodeJS application, where I do a Promise.all on a multi-cache wrap using the current version. See how I'm able to effectively do start a bunch of queries while the first one is not even completed, and hasn't got it's result back yet :

![image](https://github.com/node-cache-manager/node-cache-manager/assets/9983268/fcc393f0-739a-43be-82d0-ebc1c36b502c)

If we don't coalesce the cache calls, we end up calling the cache database multiple times for the same exact query because we are indeed calling that cache multiple times before it's completed. It's especially relevant when you use a multi-layer-cache approach and use a memory cache. You would expect to only do a single query to the database, and then serve the object from memory, but what happens is just a bunch of queries to the database.

**Comment by BryanDonovan** [comment]
Yes, the original reason for this was to handle a scenario like:

1. 10 simultaneous requests are received for user 123, which isn't cached yet.
2. Instead of trying to fetch the same key from the cache 10 times in parallel, and have each one miss and then have to fetch from the database, we instead put the requests in a "callback filler" queue, so only one request actually calls the cache and database.
3. When the single request comes back, it sends the data back in its callback and each of the other 9 callbacks.

So the whole point is to _avoid_ trying to make parallel calls to the cache or database for the same key.

**Comment by botika** [comment]
I'm not saying the database call, I'm saying the TCP request. The requests are made one by one by thread. Makes the request and makes the response, request/response. 

I may be wrong, I haven't debugged inside nodejs, but the Rust implementation works on the tokio runtime which is quite similar to the nodejs event loop, reactor pattern. According to my logs, a call enters and that call ends before continuing with the next one. Correct me if I'm wrong.

**Comment by botika** [comment]
My bad. Sorry I don't know why I thought of this. It doesn't have much to do with the tcp implementation of one and the other, you're right. https://github.com/ntex-rs/ntex/blob/996f2afdb7bf9b0a5757aa864458fddee9f1f72f/ntex/src/server/accept.rs#L408

**Comment by botika** [comment]
@Maelstrom96 My bad, you are right.

**Comment by douglascayers** [comment]
Hi @jaredwray, I see this issue has been closed but we still experience this behavior change.

As @Maelstrom96 described originally, without the "locking" or "coalescing" functionality of the `CallbackFiller` from the v4.x line, then when multiple calls to `wrap` for the same key occur close together then all those requests have their generator function invoked, which pummel databases and downstream apis.

Does the team intend to bring this functionality back or is this issue closed as "won't fix"?

Thanks!

**Comment by jaredwray** [comment]
@douglascayers - thanks for the response and at this time we do not plan to re-implement this. I am more than happy to look at a pull request around this.

**Comment by douglascayers** [comment]
> @douglascayers - thanks for the response and at this time we do not plan to re-implement this. I am more than happy to look at a pull request around this.

Thanks for the quick follow up! I'm working on porting the logic into our app, and would be happy to submit a pull request for your all's consideration for this project.

**Comment by jaredwray** [comment]
That would be great. I would also love to see some unit tests that replicate this so we know we have a workable solution.

**Comment by douglascayers** [comment]
✅ Step 1: create https://www.npmjs.com/package/promise-coalesce

⏳ Step 2: create pull request to update the `wrap` function to use the coalescing logic
